lib/music_service/services/spotify.dart

```diff
@@ -0,0 +1,201 @@
+import 'dart:convert';
+import 'dart:io';
+import 'dart:math';
+
+import 'package:http/http.dart' as http;
+import 'package:song_tinder/music_service/music_service_interface.dart';
+import 'package:song_tinder/models/song_model.dart';
+import 'package:song_tinder/models/playlist_model.dart';
+import 'package:song_tinder/models/artist_model.dart';
+import 'package:song_tinder/models/album_model.dart';
+import 'package:url_launcher/url_launcher.dart';
+
+class Spotify extends MusicServiceInterface {
+  Spotify() {
+    _initAuth();
+  }
+
+  final client = http.Client();
+
+  late String accessToken;
+  static const String _clientId = ""; // Your Spotify app clientId;
+  static const _clientSecret = ""; // Your Spotify app clientSecret;
+
+  static const String _callbackServerURL = "http://localhost:3000";
+  final Uri _initAuthURL = Uri(
+      scheme: 'https',
+      host: 'accounts.spotify.com',
+      path: '/authorize',
+      queryParameters: {
+        'client_id': _clientId,
+        'redirect_uri': _callbackServerURL,
+        'response_type': 'code',
+        'scope':
+            'ugc-image-upload user-modify-playback-state user-read-playback-state user-read-currently-playing user-follow-modify user-follow-read user-read-recently-played user-read-playback-position user-top-read playlist-read-collaborative playlist-modify-public playlist-read-private playlist-modify-private app-remote-control streaming user-read-email user-read-private user-library-modify user-library-read'
+      });
+
+  
+  Future<http.Response> _sendRequest(String path, Map<String, String>? queryParameters, bool isPost) async {
+    http.Response response;
+    if (isPost) {
+      var uri = Uri.https('api.spotify.com', '/v1$path');
+      response = await client.post(uri, headers: {'Authorization': 'Bearer ' + accessToken}, body: queryParameters);
+    }
+    else {
+      var uri = Uri.https('api.spotify.com', "/v1$path", queryParameters);
+      print(uri.toString());
+      response = await client.get(uri, headers: {'Authorization': 'Bearer ' + accessToken});
+    }
+    if (response.statusCode == 401) {
+      _renewAuth();
+      return _sendRequest(path, queryParameters, isPost);
+    }
+    if (response.statusCode == 200) {
+      return response;
+    }
+    throw HttpException("Unexpected response status: ${response.statusCode}");
```

Comment:
it's generally best to have code look somthing like this:
```dart
if (response.statusCode is error) { exit }

if (response.statusCode requires special handling) { handle special case }

return success...
```
It helps avoid nesting if else statements or making them reliant on each other, and it makes it easy to create and "is error" functions or the like and make it handle the specifics of the codes themselves while the top level function (_sendRequest in this case) can be as close to pseudo code as possible.